text/0000-cargo-test-coverage.md

```diff
@@ -0,0 +1,249 @@
+to run and obtain code coverage results for a specific set of crates to ensure confidence
+in code quality and correctness. Currently, in order to get instrumentation based code coverage,
+a Rust developer would have to either update the `RUSTFLAGS` environment variable or cargo
+manifest keys with `-C instrument-coverage`. This would automatically enable instrumentation
+of all Rust crates within the dependency graph, not just the top level crate. Instrumenting
+all crates including transitive dependencies does not help the developer ensure test coverage
+for the crates they actually want to test. This also adds unnecessary work for both the Rust
+compiler and codegen backend, `LLVM` in this case, to instrument all libraries as opposed to
+the subset a Rust developer actually cares about.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This section examines the features proposed by this RFC:
+
+## CLI option
+
+A new flag for the `cargo test` command would be added. The new flag `--coverage` would instruct Cargo to
+add the Rust compiler flag `-C instrument-coverage`, for the given crate only. This would mean that only the top-level
+crate would be instrumented and code coverage results would only run against this crate. As an example, lets take the
+following crate `foo`:
+
+```text
+/Cargo.toml
+  +-- src
+      +-- lib.rs
+```
+
+Where crate `foo` has a dependency on the `regex`:
+```toml
+[dependencies]
+regex = "*"
+```
+
+And `lib.rs` contains:
+
+```rust
+use regex::*;
+
+pub fn match_regex(re: Regex, text: &str) -> bool {
+    let Some(caps) = re.captures(text) else {
+        return false
+    };
+
+    let matches = caps.iter().filter_map(|s| s).collect::<Vec<Match>>();
+    matches.len() > 0
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn find_match() {
+        let result = match_regex(Regex::new(".*").unwrap(), "Hello World");
+        assert_eq!(result, true);
+    }
+
+    #[test]
+    fn find_no_match() {
+        let result = match_regex(Regex::new("a+").unwrap(), "Hello World");
+        assert_eq!(result, false);
+    }
+}
+```
+
+Now running `cargo test --coverage` would build the `foo` crate and instrument all libraries created by this crate.
+Next, the tests for the crate `foo` will be run and produce code coverage results for this crate only and ignore
+all code coverage for any function defined outside of this crate.
+
+The `cargo test` subcommand also supports the `--package` and `--workspace` flags. This instructs cargo to run the
+tests for those specific packages. When combined with the new `--coverage` flag, a Rust developer will be able to
+selectively choose which Rust crates will be instrumented and have tests run. Since potentially unit tests from
+multiple crates will be run, the code coverage results will include coverage results for a crate that is being
+covered by a test from another crate.
```

Comment:
Should there be a way to say "run my local tests, but also include coverage data from `foo` without running its tests"?

Reply:
Ah, I see that's under future possibilities. It seems likely to me that you would sometimes want the coverage data without the test run, so I think it makes sense to include that now.

Reply:
Will do. I'll move this from under the future possibilities section and include it now.